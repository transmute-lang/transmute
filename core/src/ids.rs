macro_rules! make_id {
    ($name:ident) => {
        #[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Default)]
        pub struct $name {
            id: usize,
        }

        impl std::fmt::Display for $name {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "{}", self.id)
            }
        }

        impl std::fmt::Debug for $name {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "{}({})", stringify!($name), self.id)
            }
        }

        impl From<usize> for $name {
            fn from(id: usize) -> Self {
                Self { id }
            }
        }

        impl From<$name> for usize {
            fn from(id: $name) -> Self {
                id.id
            }
        }

        impl std::ops::Add<usize> for $name {
            type Output = $name;

            fn add(self, rhs: usize) -> Self::Output {
                Self::from(self.id + rhs)
            }
        }

        impl std::ops::Add<usize> for &$name {
            type Output = $name;

            fn add(self, rhs: usize) -> Self::Output {
                $name::from(self.id + rhs)
            }
        }
    };
}

pub(crate) use make_id;

make_id!(IdentId);
make_id!(IdentRefId);
make_id!(ExprId);
make_id!(StmtId);
make_id!(SymbolId);
make_id!(TypeId);
make_id!(FunctionId);
make_id!(StructId);
make_id!(TypeDefId);

#[macro_export]
macro_rules! id {
    ($e:expr) => {
        <_ as Into<usize>>::into($e)
    };
}

pub use id;
