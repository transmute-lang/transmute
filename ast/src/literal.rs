use std::fmt::{Debug, Formatter};
use transmute_core::ids::IdentRefId;
use transmute_core::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Literal {
    pub kind: LiteralKind,
    pub span: Span,
}

impl Literal {
    pub fn new(kind: LiteralKind, span: Span) -> Self {
        Self { kind, span }
    }
}

#[derive(Clone, PartialEq)]
pub enum LiteralKind {
    Boolean(bool),
    Identifier(IdentRefId),
    Number(i64),
}

impl Debug for LiteralKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            LiteralKind::Boolean(v) => write!(f, "Boolean({v})"),
            LiteralKind::Identifier(v) => write!(f, "IdentRefId({v})"),
            LiteralKind::Number(v) => write!(f, "Number({v})"),
        }
    }
}
