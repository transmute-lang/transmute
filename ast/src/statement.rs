use crate::annotation::Annotation;
use crate::identifier::Identifier;
use transmute_core::ids::{ExprId, IdentRefId, InputId, StmtId, TypeDefId};
use transmute_core::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Statement {
    pub id: StmtId,
    pub kind: StatementKind,
    pub span: Span,
}

impl Statement {
    pub fn new(id: StmtId, kind: StatementKind, span: Span) -> Self {
        Self { id, kind, span }
    }

    pub fn as_namespace(&self) -> (&Identifier, InputId, &[StmtId]) {
        match &self.kind {
            StatementKind::Namespace(identifier, input_id, stmts) => (identifier, *input_id, stmts),
            _ => panic!("{:?} is not a namespace", self),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum StatementKind {
    Expression(ExprId),
    Let(Identifier, ExprId),
    Ret(Option<ExprId>, RetMode),
    LetFn(Identifier, Vec<Annotation>, Vec<Parameter>, Return, ExprId),
    Struct(Identifier, Vec<Annotation>, Vec<Field>),
    Annotation(Identifier),
    Use(Vec<IdentRefId>),
    Namespace(
        /// Namespace identifier
        Identifier,
        /// `InputId` this namespace is coming from
        InputId,
        /// Statements included in this namespace
        Vec<StmtId>,
    ),
}

/// AST only produces `Ret` with `RetMode::Explicit` but having that information in the AST nodes
/// helps in the hir transformations (see `ImplicitRetConverter::visit_expression`).
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum RetMode {
    Explicit,
    Implicit,
}

impl RetMode {
    pub fn as_str(&self) -> &'static str {
        match self {
            RetMode::Explicit => "explicit",
            RetMode::Implicit => "implicit",
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct TypeDef {
    pub kind: TypeDefKind,
    pub span: Span,
}

impl TypeDef {
    pub fn identifier_ref_id(&self) -> &[IdentRefId] {
        match &self.kind {
            TypeDefKind::Simple(ident_ref_id) => ident_ref_id,
            TypeDefKind::Array(_, _) => todo!(),
            _ => panic!("no identifier_ref_id for TypeDefKind::Dummy"),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum TypeDefKind {
    Simple(Vec<IdentRefId>),
    Array(TypeDefId, usize),
    /// A dummy type kind, inserted by the parser when the type could not be parsed
    Dummy,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Parameter {
    pub identifier: Identifier,
    pub type_def_id: TypeDefId,
    pub span: Span,
}

impl Parameter {
    pub fn new(identifier: Identifier, type_def_id: TypeDefId, span: Span) -> Self {
        Self {
            identifier,
            type_def_id,
            span,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Return {
    pub type_def_id: Option<TypeDefId>,
}

impl Return {
    pub fn none() -> Self {
        Self { type_def_id: None }
    }

    pub fn some(type_def_id: TypeDefId) -> Self {
        Self {
            type_def_id: Some(type_def_id),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Field {
    pub identifier: Identifier,
    pub type_def_id: TypeDefId,
    pub span: Span,
}

impl Field {
    pub fn new(identifier: Identifier, type_def_id: TypeDefId, span: Span) -> Self {
        Self {
            identifier,
            type_def_id,
            span,
        }
    }
}
