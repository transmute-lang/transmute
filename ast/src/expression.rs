use crate::literal::Literal;
use crate::operators::{BinaryOperator, UnaryOperator};
use std::fmt::Debug;
use transmute_core::ids::{ExprId, IdentRefId, StmtId};
use transmute_core::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Expression {
    pub id: ExprId,
    pub kind: ExpressionKind,
    pub span: Span,
}

impl Expression {
    pub fn new(id: ExprId, kind: ExpressionKind, span: Span) -> Self {
        Self { id, kind, span }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExpressionKind {
    Assignment(Target, ExprId),
    If(ExprId, ExprId, Option<ExprId>),
    Literal(Literal),
    Binary(ExprId, BinaryOperator, ExprId),
    Unary(UnaryOperator, ExprId),
    Access(ExprId, IdentRefId),
    FunctionCall(IdentRefId, Vec<ExprId>),
    While(ExprId, ExprId),
    Block(Vec<StmtId>),
    StructInstantiation(IdentRefId, Vec<(IdentRefId, ExprId)>),
    /// A dummy expression kind, inserted by the parser when the expression could not be parsed
    Dummy,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Target {
    Direct(IdentRefId),
    // The expression id is of kind ExpressionKind::Access
    Indirect(ExprId),
}
