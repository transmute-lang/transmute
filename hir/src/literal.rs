use std::fmt::{Debug, Formatter};
use transmute_ast::literal::Literal as AstLiteral;
use transmute_ast::literal::LiteralKind as AstLiteralKind;
use transmute_core::ids::IdentRefId;
use transmute_core::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Literal {
    pub kind: LiteralKind,
    pub span: Span,
}

impl Literal {
    pub fn new(kind: LiteralKind, span: Span) -> Self {
        Self { kind, span }
    }
}

impl From<AstLiteral> for Literal {
    fn from(value: AstLiteral) -> Self {
        Self {
            span: value.span.clone(),
            kind: LiteralKind::from(value.kind),
        }
    }
}

#[derive(Clone, PartialEq)]
pub enum LiteralKind {
    Boolean(bool),
    Identifier(IdentRefId),
    Number(i64),
}

impl From<AstLiteralKind> for LiteralKind {
    fn from(value: AstLiteralKind) -> Self {
        match value {
            AstLiteralKind::Boolean(b) => Self::Boolean(b),
            AstLiteralKind::Identifier(i) => Self::Identifier(i),
            AstLiteralKind::Number(n) => Self::Number(n),
        }
    }
}

impl Debug for LiteralKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            LiteralKind::Boolean(v) => write!(f, "Boolean({v})"),
            LiteralKind::Identifier(v) => write!(f, "IdentRefId({v})"),
            LiteralKind::Number(v) => write!(f, "Number({v})"),
        }
    }
}
