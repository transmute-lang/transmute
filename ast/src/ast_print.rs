use crate::expression::{Expression, ExpressionKind, Target};
use crate::literal::{Literal, LiteralKind};
use crate::statement::{Statement, StatementKind, TypeDefKind};
use std::fmt::Write;
use transmute_core::id;
use transmute_core::ids::{ExprId, IdentId, IdentRefId, StmtId, TypeDefId};

pub trait AstPrint {
    fn ast_print<W>(
        &self,
        ctx: &mut AstPrintContext<'_>,
        opts: &Options,
        f: &mut W,
    ) -> std::fmt::Result
    where
        W: Write;
}

#[derive(Default)]
pub struct Options {}

impl AstPrint for Literal {
    fn ast_print<W>(
        &self,
        ctx: &mut AstPrintContext<'_>,
        _opts: &Options,
        f: &mut W,
    ) -> std::fmt::Result
    where
        W: Write,
    {
        match &self.kind {
            LiteralKind::Boolean(b) => {
                writeln!(f, "{indent}Bool: {b}", indent = ctx.indent())
            }
            LiteralKind::Identifier(ident) => {
                writeln!(
                    f,
                    "{indent}Ident: {ident}",
                    indent = ctx.indent(),
                    ident = ctx.identifier_ref(*ident)
                )
            }
            LiteralKind::Number(n) => {
                writeln!(f, "{indent}Number: {n}", indent = ctx.indent())
            }
            LiteralKind::String(n) => {
                writeln!(f, "{indent}String: {n}", indent = ctx.indent())
            }
        }
    }
}

impl AstPrint for Expression {
    fn ast_print<W>(
        &self,
        ctx: &mut AstPrintContext<'_>,
        opts: &Options,
        f: &mut W,
    ) -> std::fmt::Result
    where
        W: Write,
    {
        if let Some(tag) = ctx.tag.take() {
            writeln!(
                f,
                "{indent}{tag}:expr {id}",
                indent = ctx.indent(),
                id = id!(self.id)
            )?;
        } else {
            writeln!(
                f,
                "{indent}expr {id}",
                indent = ctx.indent(),
                id = id!(self.id)
            )?;
        }

        ctx.next_level();
        match &self.kind {
            ExpressionKind::Assignment(Target::Direct(target), expr_id) => {
                writeln!(
                    f,
                    "{indent}Assignment target={ident}",
                    indent = ctx.indent(),
                    ident = ctx.identifier_ref(*target)
                )?;
                ctx.last = true;
                ctx.next_level();
                ctx.last = true;
                ctx.ast_print_expression(*expr_id, opts, f)?;
                ctx.prev_level();
            }
            ExpressionKind::Assignment(Target::Indirect(lhs_expr_id), rhs_expr_id) => {
                writeln!(f, "{indent}Assignment", indent = ctx.indent())?;
                ctx.last = true;
                ctx.next_level();

                ctx.tag = Some("target");
                ctx.ast_print_expression(*lhs_expr_id, opts, f)?;
                ctx.tag = Some("value");
                ctx.last = true;
                ctx.ast_print_expression(*rhs_expr_id, opts, f)?;
                ctx.prev_level();
            }
            ExpressionKind::If(cond_id, true_id, false_id) => {
                writeln!(f, "{indent}If", indent = ctx.indent())?;
                ctx.last = true;
                ctx.next_level();

                ctx.tag = Some("cond");
                ctx.ast_print_expression(*cond_id, opts, f)?;

                ctx.tag = Some("true");
                ctx.last = !false_id.is_some();
                ctx.ast_print_expression(*true_id, opts, f)?;

                if let Some(false_id) = false_id {
                    ctx.tag = Some("false");
                    ctx.last = true;
                    ctx.ast_print_expression(*false_id, opts, f)?;
                }

                ctx.prev_level();
            }
            ExpressionKind::Literal(lit) => lit.ast_print(ctx, opts, f)?,
            ExpressionKind::Binary(left_id, op, right_id) => {
                writeln!(
                    f,
                    "{indent}Binary {op}",
                    indent = ctx.indent(),
                    op = op.kind.as_str()
                )?;

                ctx.last = true;
                ctx.next_level();
                ctx.tag = Some("lhs");
                ctx.ast_print_expression(*left_id, opts, f)?;
                ctx.tag = Some("rhs");
                ctx.last = true;
                ctx.ast_print_expression(*right_id, opts, f)?;
                ctx.prev_level();
            }
            ExpressionKind::Unary(op, expr_id) => {
                writeln!(
                    f,
                    "{indent}Unary {op}",
                    indent = ctx.indent(),
                    op = op.kind.as_str()
                )?;

                ctx.next_level();
                ctx.ast_print_expression(*expr_id, opts, f)?;
                ctx.prev_level();
            }
            ExpressionKind::Access(expr_id, ident_ref_id) => {
                writeln!(f, "{indent}Access", indent = ctx.indent())?;

                ctx.last = true;
                ctx.next_level();
                ctx.ast_print_expression(*expr_id, opts, f)?;
                writeln!(
                    f,
                    "{indent}ident={ident}",
                    indent = ctx.indent(),
                    ident = ctx.identifier_ref(*ident_ref_id)
                )?;
                ctx.prev_level();
            }
            ExpressionKind::FunctionCall(expr_id, param_ids) => {
                writeln!(f, "{indent}Call", indent = ctx.indent())?;

                ctx.last = true;
                ctx.next_level();
                ctx.ast_print_expression(*expr_id, opts, f)?;

                if param_ids.is_empty() {
                    writeln!(f, "{indent}params=[]", indent = ctx.indent())?;
                } else {
                    writeln!(f, "{indent}params", indent = ctx.indent())?;
                    ctx.last = true;
                    ctx.next_level();
                    for (i, param) in param_ids.iter().enumerate() {
                        ctx.last = i + 1 == param_ids.len();
                        ctx.ast_print_expression(*param, opts, f)?;
                    }
                    ctx.prev_level();
                }

                ctx.prev_level();
            }
            ExpressionKind::While(cond_id, expr_id) => {
                writeln!(f, "{indent}While", indent = ctx.indent())?;
                ctx.last = true;
                ctx.next_level();

                ctx.tag = Some("cond");
                ctx.ast_print_expression(*cond_id, opts, f)?;
                ctx.tag = Some("body");
                ctx.last = true;
                ctx.ast_print_expression(*expr_id, opts, f)?;

                ctx.prev_level();
            }
            ExpressionKind::Block(stmt_ids) => {
                if stmt_ids.is_empty() {
                    writeln!(f, "{indent}Block=[]", indent = ctx.indent())?;
                } else {
                    writeln!(f, "{indent}Block", indent = ctx.indent())?;
                    ctx.last = true;
                    ctx.next_level();
                    for (i, stmt_id) in stmt_ids.iter().enumerate() {
                        ctx.last = i + 1 == stmt_ids.len();
                        ctx.ast_print_statement(*stmt_id, opts, f)?;
                    }
                    ctx.prev_level();
                }
            }
            ExpressionKind::StructInstantiation(ident_ref_id, fields) => {
                writeln!(
                    f,
                    "{indent}StructInstantiate name={ident}",
                    indent = ctx.indent(),
                    ident = ctx.identifier_ref(*ident_ref_id)
                )?;
                ctx.last = true;
                ctx.next_level();
                for (i, (field_ident, field_value)) in fields.iter().enumerate() {
                    writeln!(
                        f,
                        "{indent}field name={ident}",
                        indent = ctx.indent(),
                        ident = ctx.identifier_ref(*field_ident),
                    )?;
                    ctx.last = i + 1 == fields.len();
                    ctx.next_level();
                    ctx.last = true;
                    ctx.ast_print_expression(*field_value, opts, f)?;
                    ctx.prev_level();
                }
                ctx.prev_level();
            }
            ExpressionKind::ArrayInstantiation(values) => {
                writeln!(f, "{indent}ArrayInstantiate", indent = ctx.indent(),)?;
                ctx.last = true;
                ctx.next_level();
                for (i, expr_id) in values.iter().enumerate() {
                    ctx.last = i + 1 == values.len();
                    ctx.ast_print_expression(*expr_id, opts, f)?;
                }
                ctx.prev_level();
            }
            ExpressionKind::ArrayAccess(base_expr_id, index_expr_id) => {
                writeln!(f, "{indent}ArrayAccess", indent = ctx.indent(),)?;

                ctx.last = true;
                ctx.next_level();
                ctx.ast_print_expression(*base_expr_id, opts, f)?;
                ctx.last = true;
                ctx.ast_print_expression(*index_expr_id, opts, f)?;
                ctx.prev_level();
            }
            ExpressionKind::Dummy => unreachable!(),
        }
        ctx.prev_level();
        Ok(())
    }
}

impl AstPrint for Statement {
    fn ast_print<W>(
        &self,
        ctx: &mut AstPrintContext<'_>,
        opts: &Options,
        f: &mut W,
    ) -> std::fmt::Result
    where
        W: Write,
    {
        if let Some(tag) = ctx.tag.take() {
            writeln!(
                f,
                "{indent}{tag}:stmt {id}",
                indent = ctx.indent(),
                id = id!(self.id)
            )?;
        } else {
            writeln!(
                f,
                "{indent}stmt {id}",
                indent = ctx.indent(),
                id = id!(self.id)
            )?;
        }
        ctx.next_level();
        match &self.kind {
            StatementKind::Expression(expr_id) => {
                writeln!(f, "{indent}Expression", indent = ctx.indent(),)?;
                ctx.last = true;
                ctx.next_level();
                ctx.last = true;
                ctx.ast_print_expression(*expr_id, opts, f)?;
                ctx.prev_level();
            }
            StatementKind::Namespace(ident, _, stmts) => {
                writeln!(
                    f,
                    "{indent}Namespace name={ident}",
                    indent = ctx.indent(),
                    ident = ctx.identifier(ident.id)
                )?;

                ctx.last = true;
                ctx.next_level();
                if stmts.is_empty() {
                    writeln!(f, "{indent}stmts=[]", indent = ctx.indent(),)?;
                } else {
                    for (i, stmt) in stmts.iter().enumerate() {
                        ctx.last = i + 1 == stmts.len();
                        ctx.ast_print_statement(*stmt, opts, f)?;
                    }
                }
                ctx.prev_level();
            }
            StatementKind::Let(ident, expr_id) => {
                writeln!(
                    f,
                    "{indent}Let name={ident}",
                    indent = ctx.indent(),
                    ident = ctx.identifier(ident.id)
                )?;
                ctx.last = true;
                ctx.next_level();
                ctx.last = true;
                ctx.ast_print_expression(*expr_id, opts, f)?;
                ctx.prev_level();
            }
            StatementKind::Ret(Some(expr_id), _mode) => {
                writeln!(f, "{indent}Ret", indent = ctx.indent())?;
                ctx.last = true;
                ctx.next_level();
                ctx.last = true;
                ctx.ast_print_expression(*expr_id, opts, f)?;
                ctx.prev_level();
            }
            StatementKind::Ret(None, mode) => {
                writeln!(
                    f,
                    "{indent}Ret ({mode})",
                    indent = ctx.indent(),
                    mode = mode.as_str()
                )?;
            }
            StatementKind::LetFn(ident, annotations, parameters, ret_type, expr_id) => {
                writeln!(
                    f,
                    "{indent}Fn name={ident}",
                    indent = ctx.indent(),
                    ident = ctx.identifier(ident.id)
                )?;
                ctx.last = true;
                ctx.next_level();

                if annotations.is_empty() {
                    writeln!(f, "{indent}annotations=[]", indent = ctx.indent())?;
                } else {
                    writeln!(f, "{indent}annotations", indent = ctx.indent())?;
                    ctx.next_level();
                    for annotation in annotations {
                        writeln!(
                            f,
                            "{indent}{ident}",
                            indent = ctx.indent(),
                            ident = annotation
                                .ident_ref_ids
                                .iter()
                                .map(|ident_ref_id| ctx.identifier_ref(*ident_ref_id))
                                .collect::<Vec<_>>()
                                .join("."),
                        )?;
                    }
                    ctx.prev_level();
                }

                if parameters.is_empty() {
                    writeln!(f, "{indent}params=[]", indent = ctx.indent())?;
                } else {
                    writeln!(f, "{indent}params", indent = ctx.indent())?;
                    ctx.next_level();
                    for parameter in parameters.iter() {
                        write!(
                            f,
                            "{indent}{ident}: ",
                            indent = ctx.indent(),
                            ident = ctx.identifier(parameter.identifier.id),
                        )?;
                        parameter.type_def_id.ast_print(ctx, opts, f)?;
                        writeln!(f)?;
                    }
                    ctx.prev_level();
                }

                if let Some(ret_type) = &ret_type.type_def_id {
                    write!(f, "{indent}ret=", indent = ctx.indent())?;
                    ret_type.ast_print(ctx, opts, f)?;
                    writeln!(f)?;
                } else {
                    writeln!(f, "{indent}ret=[]", indent = ctx.indent())?;
                }

                ctx.tag = Some("body");
                ctx.last = true;
                ctx.ast_print_expression(*expr_id, opts, f)?;

                ctx.prev_level();
            }
            StatementKind::Struct(ident, _annotations, fields) => {
                writeln!(
                    f,
                    "{indent}Struct name={ident}",
                    indent = ctx.indent(),
                    ident = ctx.identifier(ident.id)
                )?;
                ctx.last = true;
                ctx.next_level();

                if fields.is_empty() {
                    writeln!(f, "{indent}fields=[]", indent = ctx.indent())?;
                } else {
                    for field in fields.iter() {
                        write!(
                            f,
                            "{indent}{ident}: ",
                            indent = ctx.indent(),
                            ident = ctx.identifier(field.identifier.id),
                        )?;
                        field.type_def_id.ast_print(ctx, opts, f)?;
                        writeln!(f)?;
                    }
                }
                ctx.prev_level();
            }
            StatementKind::Use(ident) => {
                writeln!(
                    f,
                    "{indent}Use {idents}",
                    indent = ctx.indent(),
                    idents = ident
                        .iter()
                        .map(|id| ctx.identifier_ref(*id))
                        .collect::<Vec<_>>()
                        .join(".")
                )?;
            }
            StatementKind::Annotation(ident) => {
                writeln!(
                    f,
                    "{indent}Annotation name={ident}",
                    indent = ctx.indent(),
                    ident = ctx.identifier(ident.id)
                )?;
            }
        }
        ctx.prev_level();
        Ok(())
    }
}

impl AstPrint for TypeDefId {
    fn ast_print<W>(
        &self,
        ctx: &mut AstPrintContext<'_>,
        opts: &Options,
        f: &mut W,
    ) -> std::fmt::Result
    where
        W: Write,
    {
        match &ctx.ast.type_defs[*self].kind {
            TypeDefKind::Simple(ident_ref_ids) => {
                write!(
                    f,
                    "{}",
                    ident_ref_ids
                        .iter()
                        .map(|ident_ref| ctx.identifier_ref(*ident_ref))
                        .collect::<Vec<_>>()
                        .join(".")
                )
            }
            TypeDefKind::Array(base, len) => {
                write!(f, "[")?;
                base.ast_print(ctx, opts, f)?;
                write!(f, "; {len}]")
            }
            TypeDefKind::Dummy => unreachable!(),
        }
    }
}

impl crate::Ast {
    pub fn ast_print<W>(&self, opts: &Options, f: &mut W) -> std::fmt::Result
    where
        W: Write,
    {
        let mut ctx = AstPrintContext {
            ast: self,
            indent: String::with_capacity(3 * 32),
            tag: None,
            last: true,
        };
        ctx.ast_print_statement(self.root, opts, f)?;
        Ok(())
    }
}

pub struct AstPrintContext<'a> {
    ast: &'a crate::Ast,
    indent: String,
    tag: Option<&'static str>,
    last: bool,
}

const NODE: &str = "+- ";
const INDENT: &str = "|  ";

// const NODE: &str = "";
// const INDENT: &str = "  ";

impl AstPrintContext<'_> {
    fn indent(&self) -> String {
        format!("{}{NODE}", self.indent)
    }

    fn next_level(&mut self) {
        if self.last {
            self.indent.push_str(&" ".repeat(INDENT.len()));
        } else {
            self.indent.push_str(INDENT);
        }
        self.last = false;
    }

    fn prev_level(&mut self) {
        for _ in 0..INDENT.len() {
            self.indent.pop();
        }
    }

    fn identifier(&self, ident_id: IdentId) -> &str {
        &self.ast.identifiers[ident_id]
    }

    fn identifier_ref(&self, ident_ref_id: IdentRefId) -> &str {
        self.identifier(self.ast.identifier_refs[ident_ref_id].ident.id)
    }

    fn ast_print_expression<W>(
        &mut self,
        expr_id: ExprId,
        opts: &Options,
        f: &mut W,
    ) -> std::fmt::Result
    where
        W: Write,
    {
        self.ast.expressions[expr_id].ast_print(self, opts, f)
    }

    fn ast_print_statement<W>(
        &mut self,
        stmt_id: StmtId,
        opts: &Options,
        f: &mut W,
    ) -> std::fmt::Result
    where
        W: Write,
    {
        self.ast.statements[stmt_id].ast_print(self, opts, f)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parse;
    use insta::assert_snapshot;
    use transmute_core::input::Input;

    macro_rules! t {
        ($name:ident) => {
            #[test]
            fn $name() {
                let ast = parse(vec![Input::Internal(
                    "internal",
                    include_str!(concat!("../../examples/", stringify!($name), ".tm")),
                )]);
                let mut s = String::new();
                ast.1
                    .unwrap()
                    .ast_print(&Options::default(), &mut s)
                    .expect("can print");
                assert_snapshot!(s);
            }
        };
    }

    t!(arrays_bounds);
    t!(arrays_if);
    t!(arrays_nested);
    t!(arrays_of_structs);
    t!(arrays_simple);
    t!(fibo_iter);
    t!(fibo_rec);
    t!(gc);
    t!(inner_function);
    t!(perimeter);
    t!(perimeter_lit_struct);
    t!(perimeter_var_struct);
    t!(print);
    t!(sibling_function);
    t!(string);
    t!(structs_anon);
    t!(structs_fn);
    t!(structs_inner);
    t!(structs_nested);
    t!(structs_of_arrays);
    t!(structs_simple);
    t!(void_fn);
}
