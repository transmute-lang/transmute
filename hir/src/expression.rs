use crate::literal::Literal;
use crate::typed::{Typed, TypedState, Untyped};
use std::fmt::Debug;
use transmute_ast::expression::Expression as AstExpression;
use transmute_ast::expression::ExpressionKind as AstExpressionKind;
use transmute_ast::expression::Target as AstTarget;
use transmute_core::ids::{ExprId, IdentRefId, StmtId, TypeId};
use transmute_core::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Expression<T>
where
    T: TypedState,
{
    pub id: ExprId,
    pub kind: ExpressionKind,
    pub span: Span,
    typed: T,
}

impl Expression<Untyped> {
    pub fn typed(self, ty: TypeId) -> Expression<Typed> {
        Expression {
            id: self.id,
            kind: self.kind,
            span: self.span,
            typed: Typed(ty),
        }
    }
}

impl From<AstExpression> for Expression<Untyped> {
    fn from(value: AstExpression) -> Self {
        Self {
            id: value.id,
            span: value.span.clone(),
            kind: match value.kind {
                AstExpressionKind::Assignment(target, expr_id) => {
                    ExpressionKind::Assignment(Target::from(target), expr_id)
                }
                AstExpressionKind::If(cond, true_branch, false_branch) => {
                    ExpressionKind::If(cond, true_branch, false_branch)
                }
                AstExpressionKind::Literal(literal) => {
                    ExpressionKind::Literal(Literal::from(literal))
                }
                AstExpressionKind::Access(expr_id, ident_ref_id) => {
                    ExpressionKind::Access(expr_id, ident_ref_id)
                }
                AstExpressionKind::FunctionCall(ident_ref_id, expr_ids) => {
                    ExpressionKind::FunctionCall(ident_ref_id, expr_ids)
                }
                AstExpressionKind::While(cond, expr_id) => ExpressionKind::While(cond, expr_id),
                AstExpressionKind::Block(expr_ids) => ExpressionKind::Block(expr_ids),
                AstExpressionKind::StructInstantiation(ident_ref_id, fields) => {
                    ExpressionKind::StructInstantiation(ident_ref_id, fields)
                }
                AstExpressionKind::ArrayInstantiation(values) => {
                    ExpressionKind::ArrayInstantiation(values)
                }
                AstExpressionKind::ArrayAccess(expr, index) => {
                    ExpressionKind::ArrayAccess(expr, index)
                }
                AstExpressionKind::Unary(_, _) => panic!("Cannot convert AstExpressionKind::Unary"),
                AstExpressionKind::Binary(_, _, _) => {
                    panic!("Cannot convert AstExpressionKind::Binary")
                }
                AstExpressionKind::Dummy => panic!("Cannot convert AstExpressionKind::Dummy"),
            },
            typed: Untyped,
        }
    }
}

impl Expression<Typed> {
    pub fn resolved_type_id(&self) -> TypeId {
        self.typed.0
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExpressionKind {
    Assignment(Target, ExprId),
    If(ExprId, ExprId, Option<ExprId>),
    Literal(Literal),
    Access(ExprId, IdentRefId),
    FunctionCall(IdentRefId, Vec<ExprId>),
    While(ExprId, ExprId),
    Block(Vec<StmtId>),
    StructInstantiation(IdentRefId, Vec<(IdentRefId, ExprId)>),
    ArrayInstantiation(Vec<ExprId>),
    ArrayAccess(ExprId, ExprId),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Target {
    Direct(IdentRefId),
    // The expression id is of kind ExpressionKind::Access
    Indirect(ExprId),
}

impl From<AstTarget> for Target {
    fn from(value: AstTarget) -> Self {
        match value {
            AstTarget::Direct(ident_ref_id) => Self::Direct(ident_ref_id),
            AstTarget::Indirect(expr_id) => Self::Indirect(expr_id),
        }
    }
}
