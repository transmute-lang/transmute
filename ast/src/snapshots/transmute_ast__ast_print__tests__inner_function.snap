---
source: ast/src/ast_print.rs
expression: s
---
+- stmt 0
   +- Namespace name=<root>
      +- stmt 1
      |  +- Use std.env.args
      +- stmt 2
      |  +- Use std.list.list_get
      +- stmt 3
      |  +- Use std.numbers.number_parse
      +- stmt 4
      |  +- Use std.numbers.print
      +- stmt 6
      |  +- Fn name=fibo
      |     +- annotations=[]
      |     +- params
      |     |  +- n: number
      |     +- ret=number
      |     +- body:expr 1
      |        +- Block
      |           +- stmt 5
      |              +- Expression
      |                 +- expr 0
      |                    +- Number: 0
      +- stmt 13
         +- Fn name=main
            +- annotations=[]
            +- params=[]
            +- ret=[]
            +- body:expr 32
               +- Block
                  +- stmt 7
                  |  +- Let name=n
                  |     +- expr 8
                  |        +- Call
                  |           +- expr 2
                  |           |  +- Ident: number_parse
                  |           +- params
                  |              +- expr 7
                  |                 +- Call
                  |                    +- expr 3
                  |                    |  +- Ident: list_get
                  |                    +- params
                  |                       +- expr 5
                  |                       |  +- Call
                  |                       |     +- expr 4
                  |                       |     |  +- Ident: args
                  |                       |     +- params=[]
                  |                       +- expr 6
                  |                          +- Number: 1
                  +- stmt 11
                  |  +- Fn name=fibo
                  |     +- annotations=[]
                  |     +- params
                  |     |  +- n: number
                  |     +- ret=number
                  |     +- body:expr 26
                  |        +- Block
                  |           +- stmt 9
                  |           |  +- Expression
                  |           |     +- expr 14
                  |           |        +- If
                  |           |           +- cond:expr 11
                  |           |           |  +- Binary <
                  |           |           |     +- lhs:expr 9
                  |           |           |     |  +- Ident: n
                  |           |           |     +- rhs:expr 10
                  |           |           |        +- Number: 2
                  |           |           +- true:expr 13
                  |           |              +- Block
                  |           |                 +- stmt 8
                  |           |                    +- Ret
                  |           |                       +- expr 12
                  |           |                          +- Ident: n
                  |           +- stmt 10
                  |              +- Expression
                  |                 +- expr 25
                  |                    +- Binary +
                  |                       +- lhs:expr 19
                  |                       |  +- Call
                  |                       |     +- expr 15
                  |                       |     |  +- Ident: fibo
                  |                       |     +- params
                  |                       |        +- expr 18
                  |                       |           +- Binary -
                  |                       |              +- lhs:expr 16
                  |                       |              |  +- Ident: n
                  |                       |              +- rhs:expr 17
                  |                       |                 +- Number: 1
                  |                       +- rhs:expr 24
                  |                          +- Call
                  |                             +- expr 20
                  |                             |  +- Ident: fibo
                  |                             +- params
                  |                                +- expr 23
                  |                                   +- Binary -
                  |                                      +- lhs:expr 21
                  |                                      |  +- Ident: n
                  |                                      +- rhs:expr 22
                  |                                         +- Number: 2
                  +- stmt 12
                     +- Expression
                        +- expr 31
                           +- Call
                              +- expr 27
                              |  +- Ident: print
                              +- params
                                 +- expr 30
                                    +- Call
                                       +- expr 28
                                       |  +- Ident: fibo
                                       +- params
                                          +- expr 29
                                             +- Ident: n
